### ISSUE TITLE
Socket spamming too much data.

### ISSUE THREAD

**Issue by oxarbitrage** [bitshares/bitshares-core#231]
Several bitshares api users and developers have complains of an issue that has not been transferred from cryptonomex to here and it is not resolved.

The initial issue is at:

https://github.com/cryptonomex/graphene/issues/540

In talks with @svk31  we found that api function "set_subscribe_callback" immediately start to send data while the correct behavior should be to start sending data after get_objects function is called. It should also receive only data from the called object. The function right now spam all object types to the websocket.

@elmato branch fixes the issue of set_subscribe_callback start sending data right away after called, branch can be found here:

https://github.com/elmato/bitshares-2/tree/no-spam-sub-function

While testing this new code i found that it is working correctly in regards to start receiving objects only after get_object function is called and after a subscription but it is still don't send data from the object requested, it send all into the websocket spamming massively it with 2.1.0 objects between others.

Here is some output using the code from branch of @elmato:

```
> alfredo@graphene:~$ wscat -c ws://localhost:8090
connected (press CTRL+C to quit)
> {"method": "call", "params": [1, "login", ["", ""]], "id": 2}
< {"id":2,"result":true}
> {"method": "call", "params": [1, "database", []], "id": 3}
< {"id":3,"result":2}
> {"method": "call", "params": [1, "history", []], "id": 4}
< {"id":4,"result":3}
> {"method": "call", "params": [2, "set_subscribe_callback", [5, true]], "id": 6}
< {"id":6,"result":null}
> {"method": "call", "params": [2, "get_objects", [["2.5.0"]]], "id": 7}
< {"id":7,"result":[{"id":"2.5.0","owner":"1.2.0","asset_type":"1.3.0","balance":4000000}]}
< {"method":"notice","params":[5,[[{"id":"2.1.0","head_block_number":367133,"head_block_id":"00059a1dd80f4deb617ccb93438bdf8ab8640013","time":"2015-10-26T12:35:00","current_witness":"1.6.25","next_maintenance_time":"2015-10-26T13:00:00","last_budget_time":"2015-10-26T12:00:00","witness_budget":75450000,"accounts_registered_this_interval":0,"recently_missed_count":0,"current_aslot":372967,"recent_slots_filled":"340282366920938463463374607431768211455","dynamic_flags":0,"last_irreversible_block_num":367112}]]]}
< {"method":"notice","params":[5,[[{"id":"2.1.0","head_block_number":367134,"head_block_id":"00059a1ef0c8e9cbe8a2cda558529bc1c6c951e7","time":"2015-10-26T12:35:03","current_witness":"1.6.22","next_maintenance_time":"2015-10-26T13:00:00","last_budget_time":"2015-10-26T12:00:00","witness_budget":75300000,"accounts_registered_this_interval":0,"recently_missed_count":0,"current_aslot":372968,"recent_slots_filled":"340282366920938463463374607431768211455","dynamic_flags":0,"last_irreversible_block_num":367113}]]]}
< {"method":"notice","params":[5,[[{"id":"2.1.0","head_block_number":367135,"head_block_id":"00059a1ff051801f0c61583ab606d1d18a0d959a","time":"2015-10-26T12:35:06","current_witness":"1.6.17","next_maintenance_time":"2015-10-26T13:00:00","last_budget_time":"2015-10-26T12:00:00","witness_budget":75150000,"accounts_registered_this_interval":0,"recently_missed_count":0,"current_aslot":372969,"recent_slots_filled":"340282366920938463463374607431768211455","dynamic_flags":0,"last_irreversible_block_num":367113}]]]}
< {"method":"notice","params":[5,[[{"id":"2.1.0","head_block_number":367136,"head_block_id":"00059a20dc8d3411d9daf51897256013ea94e419","time":"2015-10-26T12:35:09","current_witness":"1.6.21","next_maintenance_time":"2015-10-26T13:00:00","last_budget_time":"2015-10-26T12:00:00","witness_budget":75000000,"accounts_registered_this_interval":0,"recently_missed_count":0,"current_aslot":372970,"recent_slots_filled":"340282366920938463463374607431768211455","dynamic_flags":0,"last_irreversible_block_num":367114}]]]}
...
> alfredo@graphene:~$
```

As you can see from that output we asked only for objects in the 2.5.X and we received the first of them and then received 2.1.0 that we didn't requested.

The problem here we think is at the bloom filter, we are open to comments on this subject.

This non desired effect has been confirmed and replicated in 2 different nodes.

A workaround we have is to change line 272 from file:

https://github.com/elmato/bitshares-2/blob/no-spam-sub-function/libraries/app/database_api.cpp

From:
`param.false_positive_probability = 1.0/10000;`
To:
`param.false_positive_probability = 1.0/100;`

With this change made and recompilation of the witness_node  we are getting the right results as it can be seen in the following output where we ask only for 1.2.0 objects and we receive only that objects types:

```
> {"method": "call", "params": [2, "get_objects", [["1.2.0"]]], "id": 7}

< {"id":7,"result":[{"id":"1.2.0","membership_expiration_date":"1969-12-31T23:59:59","registrar":"1.2.0","referrer":"1.2.0","lifetime_referrer":"1.2.0","network_fee_percentage":2000,"lifetime_referrer_fee_percentage":8000,"referrer_rewards_percentage":0,"name":"committee-account","owner":{"weight_threshold":1,"account_auths":[],"key_auths":[],"address_auths":[]},"active":{"weight_threshold":29035,"account_auths":[["1.2.1191",2007],["1.2.22517",2539],["1.2.90742",5947],["1.2.90744",5947],["1.2.90746",5947],["1.2.90747",5947],["1.2.90748",5947],["1.2.90749",5947],["1.2.90750",5947],["1.2.90751",5947],["1.2.90752",5947]],"key_auths":[],"address_auths":[]},"options":{"memo_key":"BTS1111111111111111111111111111111114T1Anm","voting_account":"1.2.5","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"statistics":"2.6.0","whitelisting_accounts":[],"blacklisting_accounts":[],"whitelisted_accounts":[],"blacklisted_accounts":[],"owner_special_authority":[0,{}],"active_special_authority":[0,{}],"top_n_control_flags":0}]}

< {"method":"notice","params":[5,[[{"id":"1.2.0","membership_expiration_date":"1969-12-31T23:59:59","registrar":"1.2.0","referrer":"1.2.0","lifetime_referrer":"1.2.0","network_fee_percentage":2000,"lifetime_referrer_fee_percentage":8000,"referrer_rewards_percentage":0,"name":"committee-account","owner":{"weight_threshold":1,"account_auths":[],"key_auths":[],"address_auths":[]},"active":{"weight_threshold":29035,"account_auths":[["1.2.1191",2007],["1.2.22517",2539],["1.2.90742",5947],["1.2.90744",5947],["1.2.90746",5947],["1.2.90747",5947],["1.2.90748",5947],["1.2.90749",5947],["1.2.90750",5947],["1.2.90751",5947],["1.2.90752",5947]],"key_auths":[],"address_auths":[]},"options":{"memo_key":"BTS1111111111111111111111111111111114T1Anm","voting_account":"1.2.5","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"statistics":"2.6.0","whitelisting_accounts":[],"blacklisting_accounts":[],"whitelisted_accounts":[],"blacklisted_accounts":[],"owner_special_authority":[0,{}],"active_special_authority":[0,{}],"top_n_control_flags":0}]]]}

< {"method":"notice","params":[5,[[{"id":"1.2.0","membership_expiration_date":"1969-12-31T23:59:59","registrar":"1.2.0","referrer":"1.2.0","lifetime_referrer":"1.2.0","network_fee_percentage":2000,"lifetime_referrer_fee_percentage":8000,"referrer_rewards_percentage":0,"name":"committee-account","owner":{"weight_threshold":1,"account_auths":[],"key_auths":[],"address_auths":[]},"active":{"weight_threshold":29049,"account_auths":[["1.2.8891",2034],["1.2.22517",2539],["1.2.90742",5947],["1.2.90744",5947],["1.2.90746",5947],["1.2.90747",5947],["1.2.90748",5947],["1.2.90749",5947],["1.2.90750",5947],["1.2.90751",5947],["1.2.90752",5947]],"key_auths":[],"address_auths":[]},"options":{"memo_key":"BTS1111111111111111111111111111111114T1Anm","voting_account":"1.2.5","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"statistics":"2.6.0","whitelisting_accounts":[],"blacklisting_accounts":[],"whitelisted_accounts":[],"blacklisted_accounts":[],"owner_special_authority":[0,{}],"active_special_authority":[0,{}],"top_n_control_flags":0}]]]}

< {"method":"notice","params":[5,[[{"id":"1.2.0","membership_expiration_date":"1969-12-31T23:59:59","registrar":"1.2.0","referrer":"1.2.0","lifetime_referrer":"1.2.0","network_fee_percentage":2000,"lifetime_referrer_fee_percentage":8000,"referrer_rewards_percentage":0,"name":"committee-account","owner":{"weight_threshold":1,"account_auths":[],"key_auths":[],"address_auths":[]},"active":{"weight_threshold":29049,"account_auths":[["1.2.8891",2034],["1.2.22517",2539],["1.2.90742",5947],["1.2.90744",5947],["1.2.90746",5947],["1.2.90747",5947],["1.2.90748",5947],["1.2.90749",5947],["1.2.90750",5947],["1.2.90751",5947],["1.2.90752",5947]],"key_auths":[],"address_auths":[]},"options":{"memo_key":"BTS1111111111111111111111111111111114T1Anm","voting_account":"1.2.5","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"statistics":"2.6.0","whitelisting_accounts":[],"blacklisting_accounts":[],"whitelisted_accounts":[],"blacklisted_accounts":[],"owner_special_authority":[0,{}],"active_special_authority":[0,{}],"top_n_control_flags":0}]]]}
> alfredo@graphene:~$
```

This modifications are not pushed to the master repo of bitshares-core because of 2 things:
1- We don't know if the code changes made by @elmato will reverberate in the GIU or other aspects of the overalll project.
2- We are not sure why the reduce in param.false_positive_probability = 1.0/100 works and what we may be sacrificing with the second change.

We need further testing from the people that use the api calls with the websocket to make this official, make further changes or whatever needed and that is why we are opening the discussion. Developer with experience in the bloom_filter will be of great help for 2) and GUI developers will be great for 1).

We really need this feedback.

Thank you very much.

**Comment by elmato** [comment]
As @oxarbitrage pointed out, there is a trade-off in the resolution of the websocket spam.

We can:

a) Leave as it is now (don't use the bloom filter): 
   Pros:  No memory consumption per websocket session (after call to set_subscribe_callback)
   Cons: Too much spam of unnecessary events. (more bandwidth consumption / more client-side processing) 

b) Use the bloom filter:
   Pros  : Less notification events (less bandwidth consumption / less client-side processing) 
   Cons : Increase the usage of memory (bounded) per websocket session (after call to set_subscribe_callback)
   
We don't address the option to keep an array with the ids of the objects that were read because the increase in memory consumption will be unbound (DOS vector) and if we limit the number of objects that can be read+subscribed the functionality will be broken.

The fc::bloom_parameters::compute_optimal_parameters (bloom_filter.hpp) function is the responsible of calculating the table size where hashes will be stored.

pec = projected_element_count
fpp = false_positive_probability

The general formula for the table size is 
size_in_Kb = pec * 1.44 * log2(1/fpp)/8.0

(the compute_optimal_parameters function cap de results to 16Kb [configurable])

If we take 5000 objects as the average amount of objects read in a user session, the simultaneous users that can be handled per **1 Gb** of memory with different false positive probabilities (tolerance of unwanted notifications) are:

~84K  users with (pec=5000, fpp=1/5000 => size=~11Kb)
~111K users with (pec=5000, fpp=1/1000 => size=~9Kb)
~166K users with (pec=5000, fpp=1/100 => size=~6Kb)

As we tolerate more false positives (more bandwidth + more client processing) the same Gb of memory can serve more users.

I think we should just tolerate the false positives and go ahead with the bloom filter.

**Comment by svk31** [comment]
Thanks for looking into this guys. I don't foresee this causing any issues for the GUI, it always requests objects directly and doesn't depend on any data that it doesn't actually want. I've even implemented some filtering of unwanted objects recently in response to this object spamming.

**Comment by abitmore** [comment]
I guess `head_block_number` and the timestamp are being pushed by default for a reason. Most if not all clients have interest in that data. Not all data in `2.1.0` is needed by all clients though, for example `last_budget_time`.

**Comment by svk31** [comment]
I've been looking into this some more, and on the GUI side there are two dependencies on notifications of object removals, for limit orders and call orders. From my testing it seems limit order removals now get included in market subscriptions, that was not the case when I wrote that part of the GUI code so that part can now be removed. That leaves the call orders, which market subscriptions do not take into accounts since only limit_order_objects or fill_operations are checked for.

One solution could be to add a check for `call_order_object` in `on_objects_removed`, that would mean adding a get_market call to call_order_object similar to what exists for limit_order_object. If we do that we could reduce network traffic from API nodes by quite a lot, since the GUI would no longer need to be notified of object removals, which are currently a huge part of the network traffic.

**Comment by svk31** [comment]
Another issue here is that `on_objects_removed` is never called in the current code, instead all object removals get passed to `on_objects_changed` in `database::notify_changed_objects()`.

**Comment by svk31** [comment]
Here's my take on elmato's changes, modified to actually use on_objects_removed and replacing the clear_filter subscribe_callback boolean with a notify_removal boolean that lets you choose to subscribe to removals or not. The bloom filter gets cleared on every call of set_subscribe_callback in this version, I preferred to reuse the boolean in order to not break set_subscribe_callback compatibility.

https://github.com/bitshares/bitshares-core/tree/231-websocket-spam

**Comment by oxarbitrage** [comment]
@svk31 Testing it here.

**Comment by svk31** [comment]
I've been testing this a lot over the last few days and I gotta say that this bloom filter does not seem to be working the way it should be. If you log the outputs of subscribe_to_item and is_subscribed_to_item as is done in the current version of the branch, you'll clearly see that there are many cases of items that are added to the bloom filter not showing up as contained in the bloom filter later, the accounts added in get_key_references is a good example.

The bottom line is that using this bloom filter with the GUI appears to be infeasible.

**Comment by svk31** [comment]
Here's an example of what I mean: on opening a new connection from the GUI the GUI first calls get_key_references, which add a number of subscriptions to the accounts in your wallet:

![image](https://cloud.githubusercontent.com/assets/6890015/22820814/b3a5d10c-ef78-11e6-9974-9e83ccdc934b.png)

For some reason the array of accounts contains duplicates so subscribe_to_item gets called twice, but we can ignore that. Then just after this the GUI uses get_full_accounts to get the account in question:

![image](https://cloud.githubusercontent.com/assets/6890015/22820930/1fc7f96e-ef79-11e6-93f7-e24960453ac2.png)

As you can see it still returns false even though it's been subscribed to __twice__ already.

Could it be related to the object type used as input to the subscribe_to_item and is_subscribed_to_item? So subscribing to a string "1.3.X" does not equal subscribing to an account id type "1.3.X"? I'm not experienced enough in C to tell if that's the case..

**Comment by svk31** [comment]
Example of a false positive with param.false_positive_probability = 1.0/10000:

![image](https://cloud.githubusercontent.com/assets/6890015/22824995/f863627a-ef8a-11e6-933d-b5d64a4a6c84.png)

I never requested object 2.12.0, but when it changed the bloom filter says I'm subscribed to it and notifies me. Could it be because 2.0.0 is close to 2.12.0?

**Comment by elmato** [comment]
@svk31 it is not the same to call subscribe_to_item with a "1.3.10" string than with a account_id_type of value "1.3.10" since the bloom filter uses the binary representation of the values.

If you look at subscribe_to_item (template function)
the line **auto vec = fc::raw::pack(i);** is making the array of bytes that later will be inserted **_subscribe_filter.insert( vec.data(), vec.size() );** in the bloom filter.

**Comment by svk31** [comment]
Yea that's what I noticed too, and in some cases the call to subscribe_to_item is made from different methods for the same object, and somehow it doesn't register so I wondered if that was the cause.

Anyway I've kept at it and think I have something workable now. I've separated out new objects from the on_objects_changed method into a separate method, and the set_subscribe_callback boolean now lets you choose whether to receive notifications of removed or newly created objects. Those are simply required for the GUI, can't work around it. There's an issue with newly created call orders or limit orders not receiving a subscription, but I've made a workaround in the GUI that fires a get_objects call whenever it receives a newly created limit or call order that it's interested in.

**Comment by elmato** [comment]
Ok, the only thing left that i want to see is this.
When inserting we are using the output of fc::raw::pack(), but when checking we are using the **contains** function of the bloom filter (that simple interpret the data as binary) without going again through fc::raw::pack().

I think we are covered for the object_id_type type but i don't know if other may raise some conflicts.
It depends on how fc::raw::pack handles non POD types
 
```
   template<typename T>
   inline bool contains(const T& t) const
   {
      return contains(reinterpret_cast<const unsigned char*>(&t),static_cast<std::size_t>(sizeof(T)));
   }

```

But supposing that the bloom filter works fine (this one, or another implementation) i have made some changes to the object_database signals.

The **notify_changed_objects** now trigger 3 signals:

- new_objects()
- changed_objects()
- removed_objects()

Then the **database_api** handles the signals in this way:

- on_new_objects: 
   Check if the newly created objects are market objects (limit_order, call_order) and dispatch them if we are subscribed to that market. 
   **[dispatch full objects]**

- on_changed_objects:
  Check if we are subscribed to any of the changed objects and dispatch them.
  Check if the changed objects are market objects (limit_order, call_order) and dispatch them if we are subscribed to that market.
  **[dispatch full objects]** 

- on_removed_objects:
  Check if we are subscribed to any of the removed objects and dispatch them.
  Check if the removed objects are market objects (limit_order, call_order) and dispatch them if we are subscribed to that market.
  **[dispatch object id]**    

*(fill_order is handled in on_applied_block)

https://github.com/elmato/bitshares-2/commits/no-spam-sub-function

**Comment by svk31** [comment]
Ok that looks like it's the same as what I did more or less, I like the market_update refactoring, I was just copy-pasting it all over the place :)

One thing though, the GUI needs to be notified of new objects, so in my version I added the boolean I mentioned that lets you subscribe to new objects.

**Comment by oxarbitrage** [comment]
https://github.com/bitshares/bitshares-core/pull/234

Pull request is a work from @elmato adding some functionality from @svk31 branch, some small commit by me at the end and some testing.

Resuming a bit, we found that set_subscribe_callback with true as notify_remove_create parameter is working good and retunrs all objects from the node as soon as it is called.

The issue was using it with false and getting only subscriptions to requested objects.

To test this better i executed the following commands:

```
 wscat -c ws://bytemaster:supersecret@localhost:8090
 {"id":1, "method":"call", "params":["database","set_subscribe_callback",[0, false]]}
{"id":1, "method":"call", "params":["database","get_objects",[["1.2.152768"]]]}
{"id":1, "method":"call", "params":["database","get_objects",[["2.6.152768"]]]}
```

- 1.2.152768 object is my account.

- 2.6.152768 is the statistics of my account. 

My account had 2 single operations before this test. On the first call i subscribe to my account:

`{"id":1, "method":"call", "params":["database","get_objects",[["1.2.152768"]]]}`


```
{"id":1,"result":[{"id":"1.2.152768","membership_expiration_date":"1970-01-01T00:00:00","registrar":"1.2.96393","referrer":"1.2.96393","lifetime_referrer":"1.2.96
393","network_fee_percentage":2000,"lifetime_referrer_fee_percentage":3000,"referrer_rewards_percentage":0,"name":"oxarbitrage.a699","owner":{"weight_threshold":1,"
account_auths":[],"key_auths":[["BTS764HHQHfJNZ6si5fFBv7BjaJTziQeuihx8EMXVNGaGkzkHYWDv",1]],"address_auths":[]},"active":{"weight_threshold":1,"account_auths":[],"k
ey_auths":[["BTS6o6byTjfThztXDD4nhzV6tR4MpjRaLVmkfiDCiTeeJfmQNKCHb",1]],"address_auths":[]},"options":{"memo_key":"BTS6o6byTjfThztXDD4nhzV6tR4MpjRaLVmkfiDCiTeeJfmQN
KCHb","voting_account":"1.2.5","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"statistics":"2.6.152768","whitelisting_accounts":[],"blacklisting_acco
unts":[],"whitelisted_accounts":[],"blacklisted_accounts":[],"owner_special_authority":[0,{}],"active_special_authority":[0,{}],"top_n_control_flags":0}]}
```

But i realized this was not going to send me any updates so i used the statistics object from that response and subscribe to it:

{"id":1, "method":"call", "params":["database","get_objects",[["2.6.152768"]]]}

{"id":1,"result":[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.7643938","total_ops":2,"total_core_in_orders":0,"lifetime_fees_paid":0,"pending_fe
es":0,"pending_vested_fees":0}]}

Note the total_ops = 2.

Now, i turned on the last version of the windows bitshares UI, connected to my node and start to make some trading. First, i purchased BitUSD, as i confirmed the order i got output in the socket for the subscribed object:

```
< {"method":"notice","params":[0,[[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.8930151","total_ops":4,"total_core_in_orders":0,"lifetime_fees_paid
":0,"pending_fees":0,"pending_vested_fees":14676}]]]}
```

The total_ops moved from 2 to 4. This is fine as bitshares made 1 normal and 1 virtual operation for each trade i made:

http://cryptofresh.com/u/oxarbitrage.a699

Then i purchased some BitCNY and after confirmation of the order the socket outputs:

```
< {"method":"notice","params":[0,[[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.8930223","total_ops":6,"total_core_in_orders":0,"lifetime_fees_paid
":14676,"pending_fees":0,"pending_vested_fees":14676}]]]}
```

Total ops moved to 6. This could be confirming that the request for specific objects its working as expected. 

A public node was created with this branch and shared in telegram dev channel.  @Bilthon made an additional test in the node:

I did the same you described and got a "2.6.x" (account_statistics_object) type object, which didn't really give me what I want, but from it I performed another request for a "2.9.x" (account_transaction_history_object), which in turn had a reference to a "1.11.x" (operation history object), which is what finally has useful information in order to update ones account balance
 It might be possible to obtain the same information with fewer API calls, but just not being spammed with all sort of notifications and actually getting a transfer notification is a huge improvement

**Comment by svk31** [comment]
Removal notifications don't appear to work properly in this version, when I close a margin position for example I don't get any notifications.

**Comment by oxarbitrage** [comment]
Guys, i had been making a final test to this and i think it is working properly. remember the following:

{"method": "call", "params": [2, "set_subscribe_callback", [5, true]], "id": 6}

will immediately subscribe to all the objects and start sending data, this is the before merge behavior we need to keep for the light wallet.

the other option is to call set_subscribe as:

{"method": "call", "params": [2, "set_subscribe_callback", [5, false]], "id": 6}

this does not start to send data inmediatly, it need a call to get object:

{"method": "call", "params": [2, "get_objects", [["2.5.0"]]], "id": 7}

that will subscribe only to updates into 2.5.0 objects.

requesting to close issue.

**Comment by oxarbitrage** [comment]
some additional test were made to check functionality of the no spam.

here i subscribe to my account while the chain is replaying, when the account is registsred the websocket is updated:

```
connected (press CTRL+C to quit)
> {"method": "call", "params": [1, "login", ["", ""]], "id": 2}
  
< {"id":2,"result":true}
> {"method": "call", "params": [1, "database", []], "id": 3}
  
< {"id":3,"result":2}
> {"method": "call", "params": [1, "history", []], "id": 4}
  
< {"id":4,"result":3}
>  {"method": "call", "params": [2, "set_subscribe_callback", [5, false]], "id": 6}
< {"id":6,"result":null}
> {"method": "call", "params": [2, "get_objects", [["1.2.152768"]]], "id": 7}
< {"id":7,"result":[null]}

...
time pass until the replay register the account ...
...

< {"method":"notice","params":[5,[[{"id":"1.2.152768","membership_expiration_date":"1970-01-01T00:00:00","registrar":"1.2.96393","referrer":"1.2.96393","lifetime
_referrer":"1.2.96393","network_fee_percentage":2000,"lifetime_referrer_fee_percentage":3000,"referrer_rewards_percentage":0,"name":"oxarbitrage.a699","owner":{"
weight_threshold":1,"account_auths":[],"key_auths":[["BTS764HHQHfJNZ6si5fFBv7BjaJTziQeuihx8EMXVNGaGkzkHYWDv",1]],"address_auths":[]},"active":{"weight_threshold"
:1,"account_auths":[],"key_auths":[["BTS6o6byTjfThztXDD4nhzV6tR4MpjRaLVmkfiDCiTeeJfmQNKCHb",1]],"address_auths":[]},"options":{"memo_key":"BTS6o6byTjfThztXDD4nhz
V6tR4MpjRaLVmkfiDCiTeeJfmQNKCHb","voting_account":"1.2.5","num_witness":0,"num_committee":0,"votes":[],"extensions":[]},"statistics":"2.6.152768","whitelisting_a
ccounts":[],"blacklisting_accounts":[],"whitelisted_accounts":[],"blacklisted_accounts":[],"owner_special_authority":[0,{}],"active_special_authority":[0,{}],"to
p_n_control_flags":0}]]]}
> 
```

another a bit more interesting example is subscribe to statistic object of the account as:

```
> {"method": "call", "params": [1, "login", ["", ""]], "id": 2}
< {"id":2,"result":true}
> {"method": "call", "params": [1, "database", []], "id": 3}
< {"id":3,"result":2}
> {"method": "call", "params": [1, "history", []], "id": 4}
< {"id":4,"result":3}
> {"method": "call", "params": [2, "set_subscribe_callback", [5, false]], "id": 6}
< {"id":6,"result":null}
> {"method": "call", "params": [2, "get_objects", [["2.6.152768"]]], "id": 7}
< {"id":7,"result":[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.9541540","total_ops":15,"total_core_in_orders":0,"lifetime_fees_paid":88056,"pe
nding_fees":0,"pending_vested_fees":0}]}

...

as time pass and transactions are made the websocket statistic object for the account is updated. this will have the object ID of the last operation.

...

< {"method":"notice","params":[5,[[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.9883183","total_ops":16,"total_core_in_orders":0,"lifetime_fees_
paid":88056,"pending_fees":0,"pending_vested_fees":0}]]]}

... another transaction ...

< {"method":"notice","params":[5,[[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.10201113","total_ops":17,"total_core_in_orders":0,"lifetime_fees_
paid":88056,"pending_fees":0,"pending_vested_fees":0}]]]}

... another transaction ...


< {"method":"notice","params":[5,[[{"id":"2.6.152768","owner":"1.2.152768","most_recent_op":"2.9.11981633","total_ops":18,"total_core_in_orders":0,"lifetime_fees
_paid":88056,"pending_fees":0,"pending_vested_fees":0}]]]}
```
